lib/core/src/refund/chain.rs

```diff
@@ -0,0 +1,314 @@
+use std::{str::FromStr, sync::Arc};
+
+use async_trait::async_trait;
+use boltz_client::{
+    boltz::{SwapTxKind, SwapType},
+    util::secrets::Preimage,
+    Amount, BtcSwapScript, BtcSwapTx,
+};
+use log::info;
+use tokio::sync::Mutex;
+
+use crate::{
+    chain::{bitcoin::BitcoinChainService, liquid::LiquidChainService},
+    error::{PaymentError, SdkResult},
+    prelude::{ChainSwap, Config, Direction, LiquidNetwork, SwapScriptV2},
+    swapper::BoltzSwapper,
+};
+
+use super::{
+    common::{self, get_lbtc_fee_rate_sats_per_vbyte},
+    RefundHandler,
+};
+
+pub(crate) struct ChainRefundHandler<SwapperType: ?Sized> {
+    config: Config,
+    swapper: Option<Arc<SwapperType>>,
+    liquid_chain_service: Arc<Mutex<dyn LiquidChainService>>,
+    bitcoin_chain_service: Arc<Mutex<dyn BitcoinChainService>>,
+}
+
+impl ChainRefundHandler<BoltzSwapper> {
+    pub(crate) fn new(
+        config: Config,
+        swapper: Arc<BoltzSwapper>,
+        liquid_chain_service: Arc<Mutex<dyn LiquidChainService>>,
+        bitcoin_chain_service: Arc<Mutex<dyn BitcoinChainService>>,
+    ) -> Self {
+        Self {
+            config,
+            swapper: Some(swapper),
+            liquid_chain_service,
+            bitcoin_chain_service,
+        }
+    }
+
+    fn estimate_btc_refund_fees_sat(
+        &self,
+        refund_tx: &BtcSwapTx,
+        refund_keypair: &boltz_client::Keypair,
+        preimage: &Preimage,
+        fee_rate_sats_per_vbyte: f64,
+    ) -> anyhow::Result<u64> {
+        let refund_tx_size = refund_tx.size(refund_keypair, preimage).map_err(|e| {
+            anyhow::anyhow!(
+                "Could not calculate refund tx size when estimating broadcast fees: {e:?}"
+            )
+        })?;
+        Ok((refund_tx_size as f64 * fee_rate_sats_per_vbyte).ceil() as u64)
+    }
+
+    async fn new_btc_refund_tx(
```

Comment:
Are we sure we want to build the refund txs ourselves?

I'm asking because, as discussed in the previous call, I'm preparing a PR for boltz-client where the created refund tx would consider all utxos, instead of just one as it does now. With this change, I'd have to duplicate those fixes here.

On a more general note: if we duplicate the refund tx building logic, we will always be behind `boltz_client` and their fixes / changes. We will have to regularly copy-paste their fixes into our version of the tx building.

Either way is fine with me, but not duplicating this logic seems more elegant and easier for us to maintain in the long term.